pi5/board.go

```diff
@@ -159,19 +149,109 @@ func (b *pinctrlpi5) Reconfigure(
 		return err
 	}
 
-	b.mu.Lock()
-	defer b.mu.Unlock()
+	// make sure every pin has a name. We already know every pin has a pin
+	for _, c := range newConf.Pins {
+		if c.Name == "" {
+			c.Name = c.Pin
+		}
+	}
+
+	if err := b.addUserDefinedNames(newConf); err != nil {
+		return err
+	}
 
 	if err := b.reconfigurePullUpPullDowns(newConf); err != nil {
 		return err
 	}
+	if err := b.reconfigureInterrupts(newConf); err != nil {
+		return err
+	}
+	b.pinConfigs = newConf.Pins
+
+	return nil
+}
+
+// reconfigureInterrupts reconfigures the digital interrupts based on the new configuration provided.
+// It reuses existing interrupts when possible and creates new ones if necessary.
+func (b *pinctrlpi5) reconfigureInterrupts(newConf *rpiutils.Config) error {
+	// look at previous interrupt config, and see if we removed any
+	for _, oldConfig := range b.pinConfigs {
+		if oldConfig.Type != rpiutils.PinInterrupt {
+			continue
+		}
+		sameInterrupt := false
+		for _, newConfig := range newConf.Pins {
+			if newConfig.Type != rpiutils.PinInterrupt {
+				continue
+			}
+			// check if we still have this interrupt
+			if oldConfig.Name == newConfig.Name && oldConfig.Pin == newConfig.Pin {
+				sameInterrupt = true
+				break
+			}
+		}
+		// if we still have the interrupt, don't modify it
+		if sameInterrupt {
+			continue
+		}
+		// we no longer want this interrupt, so we will remove it and add back the pin's gpio functionality
+		bcom, ok := rpiutils.BroadcomPinFromHardwareLabel(oldConfig.Pin)
+		if !ok {
+			return errors.Errorf("cannot find GPIO for unknown pin: %s", oldConfig.Name)
+		}
+		// this actually removes the interrupt
+		interrupt, ok := b.interrupts[bcom]
+		if ok {
+			if err := interrupt.Close(); err != nil {
+				return err
+			}
+			delete(b.interrupts, bcom)
+		}
+
+		// add back the gpio pin to make it available to the user
+		b.gpios[bcom] = b.boardPinCtrl.CreateGpioPin(b.gpioMappings[oldConfig.Pin], rpiutils.DefaultPWMFreqHz)
+	}
+	// add any new interrupts. DigitalInterruptByName will create the interrupt only if we are not already managing it.
+	for _, newConfig := range newConf.Pins {
+		if newConfig.Type != rpiutils.PinInterrupt {
+			continue
+		}
+		if _, err := b.DigitalInterruptByName(newConfig.Name); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// record all custom pin names that the user has defined in the config for lookup.
+func (b *pinctrlpi5) addUserDefinedNames(newConf *rpiutils.Config) error {
+	nameToPin := map[string]uint{}
+	for _, pinConf := range newConf.Pins {
+		// check if the pin name matches a name we handle by default
+		_, alreadyDefined := rpiutils.BroadcomPinFromHardwareLabel(pinConf.Name)
```

Comment:
could you use this to identify if you have any new pins in the config? or would that be more messy because you just have strings of name and pin to compare